src/toil/utils/toilKill.py

```diff
@@ -25,28 +26,47 @@
 
 def main() -> None:
     parser = parser_with_common_options()
+    parser.add_argument('--force', action='store_true',
+                        help="Send SIGKILL to the leader process if local.")
     options = parser.parse_args()
     set_logging_from_options(options)
     config = Config()
     config.setOptions(options)
 
-    # Kill the pid recorded in the job store.
+    # Get the job store
     try:
         job_store = Toil.resumeJobStore(config.jobStore)
     except NoSuchJobStoreException:
         logger.error("The job store %s does not exist.", config.jobStore)
         return
 
+    # Get the leader PID
     with job_store.read_shared_file_stream("pid.log") as f:
         pid_to_kill = int(f.read().strip())
 
-    # TODO: We assume this is a local PID so we can't kill the leader if the
-    #  process is on a different machine. We also rely on the leader to be
-    #  responsive to clean up all its jobs.
+    # Check if the leader is on the same machine
+    with job_store.read_shared_file_stream("leader_node_id.log") as f:
+        leader_node_id = f.read().decode('utf-8').strip()
+    local_leader = leader_node_id == getNodeID()
 
-    try:
-        os.kill(pid_to_kill, signal.SIGTERM)
-        logger.info("Toil process %i successfully terminated.", pid_to_kill)
-    except OSError:
-        logger.error("Toil process %i could not be terminated.", pid_to_kill)
-        raise
+    if local_leader:
+        # Check if we can send signal to the leader. If not, process might be in
+        # another container so we fall back to using the kill flag through the
+        # job store.
+        try:
+            os.kill(pid_to_kill, 0)
+        except OSError:
+            local_leader = False
+
+    if local_leader:
+        try:
+            os.kill(pid_to_kill, signal.SIGKILL if options.force else signal.SIGTERM)
+            logger.info("Toil process %i successfully terminated.", pid_to_kill)
+        except OSError:
+            logger.error("Could not signal process %i. Is it still running?", pid_to_kill)
+            sys.exit(1)
+    else:
+        # Flip the flag inside the job store to signal kill
+        with job_store.write_shared_file_stream("_toil_kill_flag") as f:
```

Comment:
Suggestion: Centralize the `"_toil_kill_flag"` file name with an environment variable like `LEADER_KILL_FLAG_FILENAME` and import it.

Also consider writing the function `_write_kill_flag_file(self)` to have a `kill=True`/`kill=False` input var, and reuse here.

Reply:
Correct me if I'm wrong but I feel like using an environment variable for this might not be appropriate? If users are allowed to change this, `toil kill` and the leader might not be working with the same file and the kill won't have any effect, especially if they are running `toil kill` from a different machine.

I like the idea of reusing these functions, although I think we'd have to move them to the job store since `toil kill` won't have a Leader instance.

Reply:
I also suspect an environment variable might not really be useful here, except maybe as a way to run a workflow that itself pathologically wants to use the filename `_toil_kill_flag`.

But it would make sense to use a constant for this instead of typing the same string in two places, and also to encapsulate this in a `write_kill_flag(True/False)` and `read_kill_flag()`. I guess those would go on the `AbstractJobStore`?

Reply:
My mistake, I did not mean an environment variable, but a constant variable.  File names can change and it's just good to have it in one place